File: cfg19/E.cpp
#include <iostream>
#include <cmath>
#include <algorithm> 
#include <set>
#include <vector>
#include <unordered_map>
using namespace std;

typedef long long int ll;



int main(){
    int T;
    cin >> T;
    while(T--){
        int n, m;
        cin >> n >> m;
        int arr[n];
        for(int i = 0; i < n; i++){
            cin >> arr[i];
        }
        set<pair<int, int> > ps;
        for(int i = 0; i < m; i++){
            int x, y;
            cin >> x >> y;
            if(x > y) swap(x, y);
            pair<int, int> p1 = make_pair(x, y);
            ps.insert(p1);
        }

        // for (auto& it: ps) {
        //     cout << it.first << "," << it.second << endl;
        // }
        // cout << endl;

        unordered_map<int, int> ms;
        for(int i = 0; i < n; i++){
            ms[arr[i]]++;
        }
        // for (auto& it: ms) {
        //     cout << it.first << ": " << it.second << endl;
        // }
        // cout << endl;

        unordered_map<int, vector<int> > group;
        for(auto& it: ms){
            int k = it.first;
            int v = it.second;
            group[v].push_back(k);
        }

        // for (auto& it: group) {
        //     cout << it.first << ": ";
        //     for(int x: it.second){
        //         cout << x << ", ";
        //     }
        //     cout << endl;
        // }
        // cout << endl;

        for(auto& it: group){
            sort(it.second.begin(), it.second.end());
        }

        // for (auto& it: group) {
        //     cout << it.first << ": ";
        //     for(int x: it.second){
        //         cout << x << ", ";
        //     }
        //     cout << endl;
        // }
        // cout << endl;

        ll res = 0;
        for(auto& it1: group){
            int c1 = it1.first;
            for(auto& it2: group){
                int c2 = it2.first;
                if(c1 < c2) continue;
                int n1 = it1.second.size();
                for(int i = n1-1; i >= 0; i--){
                    int x = it1.second[i];
                    int n2 = it2.second.size();
                    for(int j = n2-1; j >= 0; j--){
                        int y = it2.second[j];
                        if(x == y) continue;
                        pair<int, int> p;
                        if(x < y){
                            p = make_pair(x, y);
                        }
                        else{
                            p = make_pair(y, x);
                        }

                        // cout << p.first << "," << p.second << endl;

                        if(!ps.count(p)){
                            // cout << "bye" << endl;

                            ll tot = (ll)(c1 + c2) * (ll)(x+y);
                            // cout << tot << endl;
                            res = max(res, tot);
                            break;
                        }
                    }
                }
            }
        }

        cout << res << '\n';
    }
}

File: cfg19/src/E.java
import java.util.*;
import java.io.*;

public class E {

	public static void main(String[] args) {
		FastScanner sc = new FastScanner();
		int T = sc.nextInt();
		while(T-->0) {
			int n = sc.nextInt();
			int m = sc.nextInt();
			int[] arr = new int[n];
			for(int i = 0; i < n; i++) {
				arr[i] = sc.nextInt();
			}
			HashSet<Pair> ps = new HashSet<>();
			for(int i = 0; i < m; i++) {
				int a = sc.nextInt();
				int b = sc.nextInt();
				ps.add(new Pair(a, b));
			}
			Multiset ms = new Multiset();
			for(int i = 0; i < n; i++) {
				ms.add(arr[i]);
			}
			HashMap<Integer, HashSet<Integer>> group = new HashMap<>();
			for(int k: ms.map.keySet()) {
				int v = ms.count(k);
				if(!group.containsKey(v)) {
					group.put(v, new HashSet<Integer>());
				}
				group.get(v).add(k);
			}
			int c = group.size();
			long[] sizes = new long[c];
			ArrayList<Integer>[] vals = new ArrayList[c];
			int ind = 0;
			for(int k: group.keySet()) {
				sizes[ind] = k;
				vals[ind] = new ArrayList<Integer>();
				for(int j: group.get(k)) {
					vals[ind].add(j);
				}
				Collections.sort(vals[ind], Collections.reverseOrder());
				ind++;
			}
			long res = 0;
			for(int i = 0; i < c; i++) {
				for(int j = i; j < c; j++) {
					int na = vals[i].size();
					int nb = vals[j].size();
					for(int u = 0; u < na; u++) {
						for(int v = 0; v < nb; v++) {
							int x = vals[i].get(u);
							int y = vals[j].get(v);
							if(x == y) continue;
							if(!ps.contains(new Pair(x, y))) {
								long tot = (sizes[i] + sizes[j]) * ((long)x + (long)y);
								res = Math.max(res, tot);
								break;
							}
						}
					}
				}
			}
			System.out.println(res);
		}

	}
	static class Pair{
		int a, b;
		public Pair(int a, int b) {
			if(a > b) {
				int aux = b; b = a; a = aux;
			}
			this.a = a; this.b = b;
		}
		@Override
		public int hashCode() {
			return a + b;
		}
		@Override
		public boolean equals(Object o) {
			Pair p = (Pair)o;
			return a == p.a && b == p.b;
		}
		public String toString() {
			return a+" "+b;
		}
	}
	static class Multiset{
		HashMap<Integer, Integer> map;
		public Multiset() {
			map = new HashMap<>();
		}
		public void add(int a) {
			if(!map.containsKey(a)) map.put(a, 1);
			else map.put(a, map.get(a)+1);
		}
		public void remove(int a) {
			map.put(a, map.get(a)-1);
			if(map.get(a) == 0) map.remove(a);
		}
		public boolean contains(int a) {
			return map.containsKey(a);
		}
		public int count(int a) {
			if(!map.containsKey(a)) return 0;
			else return map.get(a);
		}
		public boolean isEmpty() {
			return map.isEmpty();
		}
		public boolean equals(Multiset ms) {
			for(int k: map.keySet()) {
				if(count(k) != ms.count(k)) return false;
			}
			return true;
		}
		public String toString() {
			return map.toString();
		}
	}
	static class FastScanner {
		public BufferedReader reader;
		public StringTokenizer tokenizer;
		public FastScanner() {
			reader = new BufferedReader(new InputStreamReader(System.in), 32768);
			tokenizer = null;
		}
		public String next() {
			while (tokenizer == null || !tokenizer.hasMoreTokens()) {
				try {
					tokenizer = new StringTokenizer(reader.readLine());
				} catch (IOException e) {
					throw new RuntimeException(e);
				}
			}
			return tokenizer.nextToken();
		}
		public int nextInt() {
			return Integer.parseInt(next());
		}
		public long nextLong() {
			return Long.parseLong(next());
		}
		public double nextDouble() {
			return Double.parseDouble(next());
		}
		public String nextLine() {
			try {
				return reader.readLine();
			} catch(IOException e) {
				throw new RuntimeException(e);
			}
		}
	}

}
